### PR TITLE
Introduce seekret

### PR DESCRIPTION
**Pull request by jcscottiii**
This PR introduces Git-Seekret
- [x] Installs git-seekret from source
  - [x] Installs Go
  - [x] Installs LibGit2
- [x] Sets up templateDir

Setup of git-seekret per repo cases covered
- [x] After checkout / clone or repos.
- [x] For existing repos
- [x] Sets up pre-commit hook

Rules enabled
- Mandrill Username, Password
- AWS ID, Key, Secret
- New Relic license key

Screenshot of the after clone case

image

Gotchas:
To ensure reliable of installation of git-seekret there are some options that need to happen, either:
- [x]  Git-seekret needs to vendor / pin their dependencies or,
- [x] Git-seekret needs to create releases

UX of it catching an error:

``` sh
$ echo "password = 'this is super secret'" > new_file
$ git add new_file
$ git commit
Found Secrets: 1
        new_file:1
            - Metadata:
              status: test
            - Rule:
              password.password
            - Content:
              password = 'this is super secret'
```

### Before merge read https://github.com/18F/laptop/pull/69#issuecomment-266769769 👀

- [ ] Update https://github.com/18F/laptop/blob/seekret/mac#L238 to reference `master`
- [ ] Update https://github.com/18F/laptop/blob/seekret/seekrets-install#L55 to reference `master`
- #118

### CODE REVIEW

**Comment by rogeruiz** [comment]
@alain-hoang This would be a good place to have standup although it'll be a little like talking into the void. But raise any concerns, questions, comments, funny memes 😉 in here and we can triage them with other engineers intrested in helping out. 👍

**Comment by rogeruiz** [comment]
:eyes: look at the tests failure here.

**Comment by alain-hoang** [comment]
See [#83](https://github.com/18F/laptop/pull/83) for more information on these failures

**Comment by rogeruiz** [comment]
Some good ideas to bring in from @cryptofilegsa

> Some Sources of Ideas for Rules:
> 
> https://github.com/michenriksen/gitrob
> and the signature JSON file here:  https://github.com/michenriksen/gitrob/blob/master/signatures.json
> https://github.com/techgaun/github-dorks
> https://github.com/danielmiessler/SecLists/tree/master/Pattern_Matching

- [ ] Private SSH key
- [ ] SSH configuration file
- [ ] Potential cryptographic private key
- [ ] Potential cryptographic key bundle
- [ ] Web browser history file
- [ ] Any single-sign-on credential storage file
- [ ] Contains word: backup
- [ ] Contains word: dump
- [ ] Contains word: password
- [ ] Contains word: credential
- [ ] Contains word: secret
- [ ] Contains words: private, key
- [ ] Pidgin OTR private key
- [ ] Shell command history file
- [ ] MySQL client command history file
- [ ] PostgreSQL client command history file
- [ ] Eclipse configuration file
- [ ] JBoss configuration xml
- [ ] PostgreSQL password file
- [ ] Ruby IRB console history file
- [ ] Pidgin chat client account configuration file
- [ ] Hexchat/XChat IRC client server list configuration file
- [ ] Irssi IRC client configuration file
- [ ] Mutt e-mail client configuration file
- [ ] AWS CLI credentials file
- [ ] T command-line Twitter client configuration file
- [ ] OpenVPN client configuration file
- [ ] Shell configuration file
- [ ] Shell profile configuration file
- [ ] Shell command alias configuration file
- [ ] Ruby On Rails secret token configuration file
- [ ] OmniAuth configuration file
- [ ] any files containing KEK or key files for data store access in java
- [ ] Ruby On Rails database schema file
- [ ] Potential Ruby On Rails database configuration file
- [ ] Django configuration file
- [ ] PHP configuration file
- [ ] KeePass password manager database file
- [ ] 1Password password manager database file
- [ ] Apple Keychain database file
- [ ] GNOME Keyring database file
- [ ] Generic system log file
- [ ] Network traffic capture file
- [ ] SQL dump file
- [ ] Jenkins publish over SSH plugin file
- [ ] Potential Jenkins credentials file
- [ ] Apache htpasswd file
- [ ] Configuration file for auto-login process
- [ ] KDE Wallet Manager database file
- [ ] Potential MediaWiki configuration file
- [ ] Tunnelblick VPN configuration file
- [ ] Rubygems credentials file
- [ ] Potential MSBuild publish profile
- [ ] Sequel Pro MySQL database manager bookmark file
- [ ] git-credential-store helper credentials file
- [ ] Git configuration file
- [ ] Chef Knife configuration file
- [ ] Chef private key
- [ ] cPanel backup ProFTPd credentials file
- [ ] Robomongo MongoDB manager configuration file
- [ ] FileZilla FTP configuration or recent servers list
- [ ] Docker configuration file
- [ ] Environment configuration file

**Comment by monfresh** [comment]
If `Ruby On Rails database schema file` refers to something like what you would find in `db/schema.rb` in a typical Rails app, then it doesn't belong in the aforementioned list because that file is supposed to be checked into version control. When [Rails generates this file](https://github.com/rails/rails/blob/fe1f4b2ad56f010a4e9b93d547d63a15953d9dc2/activerecord/lib/active_record/schema_dumper.rb#L64), you'll notice the following comment:

```
It's strongly recommended that you check this file into your version control system.
```

Same with `Ruby On Rails secret token configuration file` and `Potential Ruby On Rails database configuration file`. Those files are typically checked into Git. It's not the files themselves that are an issue, it's whether or not they contain sensitive information that should be stored in an ENV var instead.

**Comment by rogeruiz** [comment]
Thanks @monfresh, I'm going to crosspost this in #86.

**Comment by jcscottiii** [comment]
@alain-hoang yeah i think there's no need to compile now if there's a binary.

**Comment by LinuxBozo** [comment]
Some last notes:
Before merge:
 
 - [ ] Update https://github.com/18F/laptop/blob/seekret/mac#L238 to reference `master`
 - [ ] Update https://github.com/18F/laptop/blob/seekret/seekrets-install#L55 to reference `master`

Travis build will fail once these updates are done since they reference things that don't exist in `master` yet.

File: mac [hunk 1]
```diff
@@ -234,14 +234,7 @@ elif [ ! -f "$HOME/.ssh/github_rsa.pub" ]; then
...
-# git config --global init.templatedir "$HOME/.git-secrets"
-# git config --global core.hooksPath "$HOME/.git-secrets/hooks"
-# git secrets --register-aws --global || true
-# find secret-patterns -type f -name '*.txt' -exec awk 'NF FNR==1{print ""}1' {} + > "$HOME/.git-secrets/patterns"
-# git secrets --add-provider --global -- cat "$HOME/.git-secrets/patterns" || true
-# git secrets --add --allowed --global 'github.com.*/[A-Za-z0-9]{40}' || true
-# git secrets --add --allowed --global 'sha.*[A-Za-z0-9]{40}' || true
-# git secrets --add --allowed --global 'secure:.*' || true
+# TODO: Change to master for merge
+curl -s https://raw.githubusercontent.com/18F/laptop/seekret/seekrets-install | sh -
```
Comment (monfresh) [inline mac]: Just leaving a reminder here to change the path to point to master before merging.

**Comment by rogeruiz** [comment]
Waiting on the last test results to pass then I'll be merging in #118 into #69 and then #69 into `master`. 🎉 thanks everyone! 🙇

**Comment by konklone** [comment]
How difficult would this be to extract from the laptop repository? We also would want to use this on potential Linux environments, and it'd be nice to make it reusable by others (and garner contributions from others) who aren't using our laptop configuration system.

**Comment by rogeruiz** [comment]
@konklone not very difficult at all. The main repository for `git-seekret` is forked [here](https://github.com/18F/git-seekret) and the main lib for that is forked [here](https://github.com/18F/seekret). @LinuxBozo [had tried getting Linux support](https://github.com/18F/git-seekret/blob/master/build.sh) out of the box but assumptions about [curl were made and need to be revisited](https://github.com/18F/git-seekret/issues/10). 

To extract the installation from this repo, we would just need to move a few things:

- the `git-seekret` documentation
- `seekrets-install`
- `seekret-rules/*.rule`
- the BATS tests
- modify the install script to point to a new location that isn't the 18F/laptop repository.

**Comment by rogeruiz** [comment]
Conversation moved to 18F/git-seekret#11

**Pull request merged**